### ISSUE TITLE
loss 수식 수정

### ISSUE THREAD

**Issue by kekmodel** [reinforcement-learning-kr/alpha_omok#16]
p_batch -> torch.exp(p_batch)

loss = F.mse_loss(v_batch, z_batch) + F.kl_div(p_batch, pi_batch)
에서 모델의 아웃풋인 p_batch의 값이 log_softmax 이므로 torch.exp(p_batch) 로 바꾸어야 softmax값이 됨.
pi_batch는 원래 softmax 값임.

**Comment by dnddnjs** [comment]
log_softmax를 output으로 사용한 이유가 있나요?
kl_div를 cross entropy 대신 사용하는 이유가 뭐에요?

**Comment by kekmodel** [comment]
loss가 nan이 발생하는 문제가 발생해서 검색해봤는데  pytorch 개발자가 한 말을 보고 조치한 것입니다.
<img width="769" alt="2018-05-09 4 00 05" src="https://user-images.githubusercontent.com/7728527/39777165-c56ad8a4-533d-11e8-8b6a-c79e4e54a6a4.png">
같은 맥락인데 cross entropy loss도 학습 테스트시 메모리크기 커지면 결과가 무한대로 나와버려서 수치적으로 안정하지 못한 것 같아 cross entropy와 동일한 결과를 내는 kl_div로 바꿨습니다.

**Comment by kekmodel** [comment]
https://github.com/pytorch/ELF/blob/59c49d101999c91411cfbfddceaff59ee2a79d90/src_py/elfgames/go/df_model3.py#L282
위의 코드를 참고해서 torch.exp()를 모델의 output 단에서 변환해서 main 코드에서는 안해줘도 되도록 다시 수정하였습니다.

**Comment by Kyushik** [comment]
이렇게 수정을 해야 확률로 나온다고 하셨죠? 참고하신 코드는.. 워 파이토치 공식인가보네요 ㅋㅋㅋㅋ

**Comment by kekmodel** [comment]
네 ㅋㅋㅋ 위에 저 사람도 파이토치 개발자 입니다. 되도록 신뢰할만한 레퍼런스를 참고하는 편입니다. :-)

**Comment by Kyushik** [comment]
혹시 푸시도 하셨나요??

**Comment by kekmodel** [comment]
제 브랜치에 했어요

**Comment by Kyushik** [comment]
이 내용은 다들 공통적으로 수정될 필요가 있는 내용 아닐까요?? 그럼 머지 해야될것 같지만 일단 웅원님 의견도 들어보시지요 ㅎㅎ 혹시 성능 개선 생기면 이야기해주세요~~ ㅎㅎ

**Comment by kekmodel** [comment]
https://github.com/pytorch/ELF/blob/59c49d101999c91411cfbfddceaff59ee2a79d90/src_py/elfgames/go/df_model3.py#L282
를 마찬가지로 참고해서 relu의 inplace 옵션도 일단 삭제해보았습니다.

**Comment by kekmodel** [comment]
loss를 크로스엔트로피로 바꿨는데 오류가 나지 않아서 수정하였습니다!
그런데 loss를 바꾸니까 그래프가 안그려지는 오류가 나서 일단 주석 처리하였습니다. 규식님이 봐주시면 감사하겠습니다.
또한 학습 전 인데도 pi가 너무 확정적이라는 규식님의 말을 듣고 pi를 "softmax -> 일반 확률"로 바꾸어서 테스트 중입니다.
